playground/imgui/imgui_impl_impeller.cc

```diff
@@ -0,0 +1,226 @@
+#include "imgui_impl_impeller.h"
+
+#include <algorithm>
+#include <climits>
+#include <memory>
+#include <vector>
+
+#include "imgui_raster.frag.h"
+#include "imgui_raster.vert.h"
+#include "third_party/imgui/imgui.h"
+
+#include "impeller/geometry/matrix.h"
+#include "impeller/geometry/point.h"
+#include "impeller/geometry/rect.h"
+#include "impeller/geometry/size.h"
+#include "impeller/renderer/allocator.h"
+#include "impeller/renderer/command.h"
+#include "impeller/renderer/context.h"
+#include "impeller/renderer/formats.h"
+#include "impeller/renderer/pipeline_builder.h"
+#include "impeller/renderer/pipeline_library.h"
+#include "impeller/renderer/range.h"
+#include "impeller/renderer/render_pass.h"
+#include "impeller/renderer/sampler.h"
+#include "impeller/renderer/sampler_library.h"
+#include "impeller/renderer/texture.h"
+#include "impeller/renderer/texture_descriptor.h"
+#include "impeller/renderer/vertex_buffer.h"
+
+struct ImGui_ImplImpeller_Data {
+  std::shared_ptr<impeller::Context> context;
+  std::shared_ptr<impeller::Texture> font_texture;
+  std::shared_ptr<impeller::Pipeline> pipeline;
+  std::shared_ptr<const impeller::Sampler> sampler;
+};
+
+static ImGui_ImplImpeller_Data* ImGui_ImplImpeller_GetBackendData() {
+  return ImGui::GetCurrentContext()
+             ? static_cast<ImGui_ImplImpeller_Data*>(
+                   ImGui::GetIO().BackendRendererUserData)
+             : nullptr;
+}
+
+bool ImGui_ImplImpeller_Init(std::shared_ptr<impeller::Context> context) {
+  ImGuiIO& io = ImGui::GetIO();
+  IM_ASSERT(io.BackendRendererUserData == nullptr &&
+            "Already initialized a renderer backend!");
+
+  // Setup backend capabilities flags
+  auto* bd = new ImGui_ImplImpeller_Data();
+  io.BackendRendererUserData = reinterpret_cast<void*>(bd);
+  io.BackendRendererName = "imgui_impl_impeller";
+  io.BackendFlags |=
+      ImGuiBackendFlags_RendererHasVtxOffset;  // We can honor the
+                                               // ImDrawCmd::VtxOffset field,
+                                               // allowing for large meshes.
+
+  bd->context = context;
+
+  // Generate/upload the font atlas.
+  {
+    unsigned char* pixels;
+    int width, height;
+    io.Fonts->GetTexDataAsRGBA32(&pixels, &width, &height);
+
+    auto texture_descriptor = impeller::TextureDescriptor{};
+    texture_descriptor.format = impeller::PixelFormat::kR8G8B8A8UNormInt;
+    texture_descriptor.size = {width, height};
+    texture_descriptor.mip_count = 1u;
+
+    bd->font_texture = context->GetPermanentsAllocator()->CreateTexture(
+        impeller::StorageMode::kHostVisible, texture_descriptor);
+    IM_ASSERT(bd->font_texture != nullptr &&
+              "Could not allocate ImGui font texture.");
+
+    bool uploaded = bd->font_texture->SetContents(pixels, width * height * 4);
```

Comment:
Please use `TextureDescriptor::GetSizeOfBaseMipLevel`. Would be less brittle in case of format changes.

Reply:
Done.